modules/drivers/druid-jdbc/src/metabase/driver/druid_jdbc.clj

```diff
@@ -0,0 +1,183 @@
+   database-type
+   :type/*))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:druid-jdbc Types/TIMESTAMP]
+  [_ ^ResultSet rs _ ^long i]
+  (fn []
+    (t/instant (.getObject rs i))))
+
+;; Druid's COMPLEX<...> types are encoded as JDBC's other -- 1111. Values are rendered as string.
+(defmethod sql-jdbc.execute/read-column-thunk [:druid-jdbc 1111]
+  [_ ^ResultSet rs _ ^long i]
+  (fn []
+    (let [o (.getObject rs i)]
+      (cond-> o
+        (string? o) (str/replace  #"^\"|\"$" "")))))
+
+(defn- date-trunc [unit expr] [:date_trunc (h2x/literal unit) expr])
+
+(defmethod sql.qp/date [:druid-jdbc :default] [_ _ expr] expr)
+(defmethod sql.qp/date [:druid-jdbc :minute] [_ _ expr] (date-trunc :minute expr))
+(defmethod sql.qp/date [:druid-jdbc :hour] [_ _ expr] (date-trunc :hour expr))
+(defmethod sql.qp/date [:druid-jdbc :day] [_ _ expr] (date-trunc :day expr))
+(defmethod sql.qp/date [:druid-jdbc :week] [_ _ expr] (date-trunc :week expr))
+(defmethod sql.qp/date [:druid-jdbc :month] [_ _ expr] (date-trunc :month expr))
+(defmethod sql.qp/date [:druid-jdbc :quarter] [_ _ expr] (date-trunc :quarter expr))
+(defmethod sql.qp/date [:druid-jdbc :year] [_ _ expr] (date-trunc :year expr))
+
+(defn- time-extract [unit expr] [:time_extract expr (h2x/literal unit)])
+
+(defmethod sql.qp/date [:druid-jdbc :minute-of-hour] [_ _ expr] (time-extract :minute expr))
+(defmethod sql.qp/date [:druid-jdbc :hour-of-day] [_ _ expr] (time-extract :hour expr))
+(defmethod sql.qp/date [:druid-jdbc :day-of-week] [_ _ expr] (time-extract :dow expr))
+(defmethod sql.qp/date [:druid-jdbc :day-of-month] [_ _ expr] (time-extract :day expr))
+(defmethod sql.qp/date [:druid-jdbc :day-of-year] [_ _ expr] (time-extract :doy expr))
+(defmethod sql.qp/date [:druid-jdbc :week-of-year] [_ _ expr] (time-extract :week expr))
+(defmethod sql.qp/date [:druid-jdbc :month-of-year] [_ _ expr] (time-extract :month expr))
+(defmethod sql.qp/date [:druid-jdbc :quarter-of-year] [_ _ expr] (time-extract :quarter expr))
+
+(defmethod sql.qp/current-datetime-honeysql-form :druid-jdbc
+  [_]
+  [:raw "CURRENT_TIMESTAMP"])
+
+(defmethod sql.qp/add-interval-honeysql-form :druid-jdbc
+  [_ hsql-form amount unit]
+  [:TIMESTAMPADD (h2x/identifier :type-name unit) (h2x/->integer amount) hsql-form])
+
+;; Query used for field values computation is not compatible with COMPLEX<...> types.
+(defmethod driver/field-values-compatible? :druid-jdbc
+  [_driver {:keys [database_type] :as _field}]
+  (not (re-find #"COMPLEX<" (str database_type))))
+
+(defmethod sql-jdbc.conn/data-source-name :druid-jdbc
+  [_driver {:keys [host port] :as _details}]
+  (format "druid-%s-%s"
+          ;; remove protocol
+          (or (re-find #"(?<=\w://).*" host)
+              host)
+          port))
+
+(defmethod sql-jdbc.execute/set-parameter [:druid-jdbc ZonedDateTime]
+  [driver ps i t]
+  (sql-jdbc.execute/set-parameter driver ps i (t/format "yyyy-MM-dd HH:mm:ss" t)))
```

Comment:
You could probably just convert this to an offset date time like `(t/offset-date-time t)` and then call `set-parameter` with that.

Reply:
That's preferable because otherwise this format string is dropping fractions of seconds which could theoretically mean

`[:= <my-timestamp-column> "2024-05-02T15:01:00.123-07:00"]` 

wouldn't work even if the value was exactly `2024-05-02T15:01:00.123-07:00`, because you'd be generating something like 

```sql
WHERE col = ? -- 2024-05-02T15:01:00-07:00 (missing the .123)
```

Reply:
On a side not I've sort of came to the conclusion that passing in `java.time` classes as separate parameters is unnecessary and buggy in a lot of JDBC drivers, I switched several drivers recently to just passing these in as timestamp literals e.g. something like 

```sql
-- Redshift
WHERE col = timestamp with time zone '2024-05-02 15:01:00.123 -07:00'
```

```sql
-- Snowflake
WHERE col = '2024-05-02 15:01:00.123 -0700'::timestamp_tz
```

Not only is this less buggy, but it's way easier to debug when things go wrong.

If things are working as is you don't need to change it but if you're running into failing timezone tests or whatever you might want to try it. Take a look at the Snowflake, Vertica, or Redshift drivers for examples of code that uses `sql.qp/->honeysql` for the `java.time` types.

If you implement things this way then you don't need `unprepare-value` methods at all

Reply:
To mitigate the described problem I've added `.SSS` suffix to the format string. Problem with the Druid is that Avatica does not implement `.setObject` with `targetSqlType` argument for various date/time classes. I've tried what you've described, but got "not implemented" error returned from Avatica driver.

Reply:
I forgot to mention it but we have `u/format-sql` which might actually format things how you want without needing to specify your own format string

Reply:
I was more suggesting that you implement `->honeysql` and splice `timestamp with time zone '2024-05-02 15:01:00.123 -07:00'` directly into the generated SQL, i.e. avoid `?` parameters altogether for temporal values. But if you have it working now we can worry about that stuff later